File: aoc/2015/aoc1.c
#include <stdlib.h>
#include <stdio.h>

int main(void)
{
    int ch, floor = 0, been_basement = 0;
    while ((ch = getchar()) != EOF)
    {
        switch (ch)
        {
        case '(':   ++floor;    break;
        case ')':   --floor;    break;
        default:
            printf("Warning: unexpected character in input: %c\n", ch);
            break;
        }

        if (!been_basement && floor == -1)
        {
            printf("Note: first entered basement at step %ld\n", ftell(stdin));
            been_basement = 1;
        }
    }

    if (ferror(stdin))
    {
        perror("Cannot read input");
        return EXIT_FAILURE;
    }

    printf("Floor: %d\n", floor);

    return EXIT_SUCCESS;
}

File: aoc/2015/rs/aoc1.rs
use std::io::{self, Read};

type ElevatorState = (i32, Option<usize>);

fn main() {
	let bytes = io::stdin().bytes()
		.map(|result| result.expect("cannot read instructions"));

		let (floor, basement_time) = run_elevator(bytes);

		println!("Ended up at floor {}.", floor);
		match basement_time {
			None => println!("Never entered the basement."),
			Some(t) => println!("Entered basement at step {}.", t)
		};
}

fn run_elevator(data: impl Iterator<Item = u8>) -> ElevatorState {
	let initial_state: ElevatorState = (0, None);

	data.map(char::from)
		.enumerate()
		.fold(initial_state, exec_instruction)
}

fn exec_instruction(
	(current_floor, basement_time): ElevatorState,
	(idx, instruction): (usize, char)
) -> ElevatorState {
	let new_floor = current_floor + match instruction {
		'(' => 1,
		')' => -1,
		_ => panic!("invalid character in input")
	};

	match basement_time {
		None if new_floor < 0 => (new_floor, Some(idx + 1)),
		_ => (new_floor, basement_time)
	}
}

#[cfg(test)]
fn run_test(input: &str, expected: ElevatorState) {
	let result = run_elevator(input.bytes());
	assert_eq!(result, expected);
}
#[test]
fn test_zero_1() { run_test("(())", (0, None)); }
